src/diffusers/models/modeling_utils.py

```diff
@@ -314,17 +328,64 @@ def save_pretrained(
-                state_dict, Path(save_directory, weights_name).as_posix(), metadata={"format": "pt"}
-            )
-        else:
-            torch.save(state_dict, Path(save_directory, weights_name).as_posix())
+        state_dict_split = split_torch_state_dict_into_shards(
+            state_dict, max_shard_size=max_shard_size, filename_pattern=weights_name
+        )
+        # Clean the folder from a previous save
+        for filename in os.listdir(save_directory):
+            full_filename = os.path.join(save_directory, filename)
+            # If we have a shard file that is not going to be replaced, we delete it, but only from the main process
+            # in distributed settings to avoid race conditions.
+            weights_no_suffix = weights_name.replace(".bin", "").replace(".safetensors", "")
+
+            # make sure that file to be deleted matches format of sharded file, e.g. pytorch_model-00001-of-00005
+            filename_no_suffix = filename.replace(".bin", "").replace(".safetensors", "")
+            reg = re.compile(r"(.*?)-\d{5}-of-\d{5}")
+
+            if (
+                filename.startswith(weights_no_suffix)
+                and os.path.isfile(full_filename)
+                and filename not in state_dict_split.filename_to_tensors.keys()
+                and is_main_process
+                and reg.fullmatch(filename_no_suffix) is not None
+            ):
+                os.remove(full_filename)
 
-        logger.info(f"Model weights saved in {Path(save_directory, weights_name).as_posix()}")
+        # Save the model
+        for filename, tensors in state_dict_split.filename_to_tensors.items():
+            shard = {tensor: state_dict[tensor] for tensor in tensors}
+            if safe_serialization:
+                # At some point we will need to deal better with save_function (used for TPU and other distributed
+                # joyfulness), but for now this enough.
+                safetensors.torch.save_file(shard, os.path.join(save_directory, filename), metadata={"format": "pt"})
+            else:
+                torch.save(shard, Path(save_directory, filename).as_posix())
+
+        index = None
+        if state_dict_split.is_sharded:
+            index = {
+                "metadata": state_dict_split.metadata,
+                "weight_map": state_dict_split.tensor_to_filename,
+            }
+
+        if index is None:
+            path_to_weights = Path(save_directory, weights_name).as_posix()
+            logger.info(f"Model weights saved in {path_to_weights}")
+        else:
+            save_index_file = SAFE_WEIGHTS_INDEX_NAME if safe_serialization else WEIGHTS_INDEX_NAME
+            save_index_file = Path(save_directory, _add_variant(save_index_file, variant)).as_posix()
+            # Save the index as well
+            with open(save_index_file, "w", encoding="utf-8") as f:
+                content = json.dumps(index, indent=2, sort_keys=True) + "\n"
+                f.write(content)
```

Comment:
(nit) Since you rely on `Path` you can do:

```py
save_index_file.write_text(
    json.dumps(index, indent=2, sort_keys=True) + "\n"
)
```

This is less verbose than `with open(...): ....`

Reply:
I have changed that to `os.path.join()` when using local paths to be consistent throughout.